include/votca/xtp/cudapipeline.h

```diff
@@ -58,7 +58,8 @@ class CudaPipeline {
   CudaPipeline &operator=(const CudaPipeline &) = delete;
 
   // C= A*b.asDiagonal()
-  void diag_gemm(const CudaMatrix &A, const CudaMatrix &b, CudaMatrix &C) const;
+  template <class M>
+  void diag_gemm(const M &A, const CudaMatrix &b, CudaMatrix &C) const;
```

Comment:
yes, it can take `cudamatrix.transpose()` or `cudamatrix.block()` which are other classes.

Reply:
that is also the reason for the template specialization furtherdown because we have to make a copy of the `transpose()` object to pass it then as a `const T&`, it only copies the `transpose() object, which holds a reference to the underlying cudamatrix` so this does not copy the matrix but only the wrapper.

Reply:
Okay I didn't know that, thanks for the clarification.